pandas/tests/frame/test_ufunc.py

```diff
@@ -60,6 +62,46 @@ def test_binary_input_dispatch_binop(dtype):
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize(
+    "func,arg,expected",
+    [
+        (np.add, 1, [2, 3, 4, 5]),
+        (
+            partial(np.add, where=[[False, True], [True, False]]),
+            np.array([[1, 1], [1, 1]]),
+            [0, 3, 4, 0],
+        ),
+        (np.power, np.array([[1, 1], [2, 2]]), [1, 2, 9, 16]),
+        (np.subtract, 2, [-1, 0, 1, 2]),
+        (
+            partial(np.negative, where=np.array([[False, True], [True, False]])),
+            None,
+            [0, -2, -3, 0],
+        ),
+    ],
+)
+def test_ufunc_passes_args(func, arg, expected, request):
+    # GH#40662
+    arr = np.array([[1, 2], [3, 4]])
+    df = pd.DataFrame(arr)
+    result_inplace = np.zeros_like(arr)
+    # 1-argument ufunc
+    if arg is None:
+        mark = pytest.mark.xfail(
+            reason="Result becomes transposed with block-wise apply of ufunc, #40878"
+        )
+        request.node.add_marker(mark)
+        result = func(df, out=result_inplace)
+    else:
+        result = func(df, arg, out=result_inplace)
+
+    expected = np.array(expected).reshape(2, 2)
+    tm.assert_numpy_array_equal(result_inplace, expected)
```

Comment:
do we have any documentation that out is actually a ndarray? this is a very strange result. At the very least document this, and let's open an issue. This should work with out=DataFrame, or simply raise (preferred)

Reply:
yep. I thought this was strange. https://github.com/pandas-dev/pandas/issues/40662#issuecomment-817307177

should we defer to 1.2.5 to allow more discussion? or put 1.2.4 release on hold for a day or two?

Reply:
yeah let's defer this. i dont' think this is the correct behavior and we should actually fix it (which may mean that we simply do this for 1.3)

Reply:
moving to 1.2.5 (if we do it)

Reply:
> This should work with out=DataFrame, or simply raise (preferred)

This does raise when passing a DataFrame to `out` (as numpy expects an array as `out` argument), but what is being tested here is the return value of the ufunc *in case of* passing an array to `out`. In that case, `out` is also being returned, and since `out` is an ndarray, the return value is also an ndarray. 

(to me this seems correct behaviour, and thus I don't think this needs to hold up merging this for the release)

Reply:
the return type is a DataFrame when out is specified and this has been added to the test

```
 expected = pd.DataFrame(expected)
 tm.assert_frame_equal(result, expected)
```

Reply:
the very fact that you can specify out as a ndarray while the input is a DataFrame is just very strange behavior and should simply not be allowed. This needs fixing and much more testing.

Reply:
> the return type is a DataFrame when out is specified and this has been added to the test

Whoops, sorry, was testing in the wrong env. Indeed, the return value is properly wrapped in a DataFrame. Indeed, now it correctly returns a DataFrame while the `out` ndarray *is* updated

> the very fact that you can specify out as a ndarray while the input is a DataFrame is just very strange behavior and should simply not be allowed. This needs fixing and much more testing.

Before DataFrame implemented this itself (with `__array_ufunc__`), any input to a numpy ufunc was coerced to an array, and `out` could also be an ndarray. This is longstanding behaviour (and IMO also fine that this works). 

If we want to change this, it can be deprecated, but for now fixing this severe regression (you silently get different values) seems the best thing to do. 
(the fix is also tiny, it's just passing through the `kwargs`, which was an oversight that this was not done before)

And note this is not only about `out`, but also about other kwargs being ignored.

Reply:
> And note this is not only about `out`, but also about other kwargs being ignored.

that was the response to my comment in #40662 https://github.com/pandas-dev/pandas/issues/40662#issuecomment-817310283 so was happy to approve and I am happy to include in 1.2.4 (but I still also think the out behavior is strange)

Reply:
ok i guess ok for 1.2.4 as this fixes the regression. but this is REALLY strange and we need to fix this (e.g. out being an ndarray), this is just plain wrong, yes this is a numpy ufunc but is so unexpected, we can simply raise an error if out is not a DataFrame I think would make this way better.

Reply:
on 2nd thought, it think this should be fixed before releasing.

Reply:
so again let's not do a last minute PR , not reason.

Reply:
Jeff, this is a tiny bugfix (look at the diff, it's just passing through `**kwargs`), that fixes a quite serious regression (you silently get wrong values when passing any additional keyword argument). 
The behavior you are objecting to is a long-standing behaviour, that we can discuss separately, but is not specific to this PR (this PR is *not* only for `out`, but fixes the regression for any keyword argument, such as `where`)

Reply:
> on 2nd thought, it think this should be fixed before releasing.

What needs to be fixed? 
If you want to change how we handle the `out` keyword (eg raising an error if it isn't a DataFrame), that can go through a deprecation cycle and thus is for another PR targeting 1.3 (and if so, please open an issue about it)

Reply:
I disagree. This is PR enables this itself by passing thru the out keyword. You say its a bug fix, but this a problem. Releasing this PR is enabling this terrible behavior.

Reply:
This "terrible" behaviour is *long-standing* (and IMO correct) behaviour on which users rely. Again, if we want to change this, we can discuss it and it can go through a deprecation cycle. But for now let's fix the regression.

Reply:
and would make DataFrame consistent with Series.. on master..

```
>>> foo = np.array([1, 2, 3, 4])
>>> foo
array([1, 2, 3, 4])
>>>
>>> ser = pd.Series(foo)
>>> ser
0    1
1    2
2    3
3    4
dtype: int32
>>>
>>> bar = np.zeros_like(ser)
>>> bar
array([0, 0, 0, 0])
>>>
>>> np.add(ser, 1, out=bar)
0    2
1    3
2    4
3    5
dtype: int32
>>>
>>> bar
array([2, 3, 4, 5])
>>>
```

Reply:
@jreback do you have time to hop on a quick call to resolve this dispute?

Reply:
> This "terrible" behaviour is long-standing (and IMO correct) behaviour on which users rely. 

how is this correct in any way? this was a bug from the original impl.

Reply:
That was not a bug, that's how numpy functions work: they coerce array-like input to arrays. So whether one of the arguments was a DataFrame vs an ndarray, did not have any effect on allowing an `out` argument or not

Reply:
ok fine. how about a follow to make this *really* clear (in docs) on master. I would however also deprecate / remove this as it not intuitive at all (e.g. we do not have an out argument anywhere else)

Reply:
> e.g. we do not have an out argument anywhere else

This keyword is not in a pandas function or method, but in a NumPy function (which has `out` in many places)

Reply:
i know, but this is *very* confusing if someone is doing this (and doesnt; realize it).